add Time-dependent DFT

EBB2675 opened the pull request:
## Purpose
I would like to add a general description for Time dependent Density functional theory.

## Scope

**Included** (_What is intentionally part of this PR?_)

- TDDFT model method description, that extends `ExcitedStateMethodology`, and reuses existing components such as `XCFunctional` and `Photon`

**Out of scope** (_What is explicitly not included (to avoid scope creep)?_)

- I haven't added any tests, as this is a highly small and straightforward PR. 

## Status

_Select one_

- [ ] Draft / In progress (not ready for full review)
- [x] Ready for review
- [ ] Blocked (explain below)

## Dependencies / Blockers

_List anything that may block progress or review_

- [x] None.
- [ ] Related PRs:
- [ ] Open design questions:
- [ ] External dependencies (people, decisions, data, releases):

## Testing / Validation

_How was this change validated?_

- [ ] Unit tests
- [ ] Integration tests
- [x] Manual testing (brief description):
- [ ] Not applicable (explain why):

- I haven't added any tests, as this is a highly small and straightforward PR. 

## Linked Issues / Context

- Fixes #
- Related to #
- Follow-up issues (if known):

## Reviewer Notes

We (me and Denis) aim to test this implementation soon rigorously in multiple parsers/flavours. 
There might of course be minor modifications needed in the future, in the light of our experience in parser dev.

EBB2675 committed:
add TDDFT

EBB2675 committed:
add solver, improve docstrings, add references

JFRudzinski commented on src/nomad_simulations/schema_packages/model_method.py:
```diff
@@ -1345,6 +1345,149 @@ class BSE(ExcitedStateMethodology):
+
+        self.name = 'TDDFT'
+
+        # Default to linear_response if not specified; infer real_time from solver hint
```
I would not default to filling in something that isn't parsed...I now wonder if there are other instances of this?

EBB2675 replied:
there are indeed

EBB2675 replied:
removed from current PR :+1:

EBB2675 committed:
remove default assignment of TDDFT.type

kubanmar commented on src/nomad_simulations/schema_packages/model_method.py:
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
+            'long_range_corrected',
+        ),
+        description="""
+        This field describes the effective XC response behaviour assumed in the TDDFT calculation, 
```
I would maybe avoid using the words "This field" or similar, for consistency

kubanmar commented on src/nomad_simulations/schema_packages/model_method.py:
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
+        """,
+    )
+
+    field_ref = Quantity(
```
"external_field_ref" ?

JosePizarro3 replied:
or `field_polarization_ref`

kubanmar commented on src/nomad_simulations/schema_packages/model_method.py:
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
+        description='External field / polarization used to drive the response or propagation.',
+    )
+
+    target = Quantity(
```
"target_quantity" or "target_property" ?

JosePizarro3 replied:
voting for `target_property` or `target_response`

kubanmar commented on src/nomad_simulations/schema_packages/model_method.py:
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
+                ):
+                    self.kernel = 'hybrid'
+            except Exception:
+                logger.warning('Could not derive TDDFT.kernel from xc_kernel_ref.')
```
This looks like if `xc_kernel_ref.global_exact_exchange` is smaller or equal 0, `kernel` will not be set. In that case we just don"t set it, or should we log a warning?

EBB2675 replied:
if `xc_kernel_ref.global_exact_exchange` is equal to 0, there are multiple options for this non hybrid `kernel`, that is simulation dependent, so not setting it here sounded appropriate

warning or info fits better :+1: will do

EBB2675 replied:
or maybe i will just avoid setting the kernel at the schema level alltogether for now

kubanmar reviewed (approved):
Looks good to me, I added some minor comments. We will see if quantities are missing when we apply it to the parsers.

ndaelman-hu commented on src/nomad_simulations/schema_packages/model_method.py:
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
+            'ALDA',
+            'AGGA',
+            'hybrid',
+            'range_separated',
```
Isn't this covered by the XC functional, already?

ndaelman-hu commented on src/nomad_simulations/schema_packages/model_method.py:
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
+    is_tamm_dancoff = Quantity(
+        type=bool,
+        description='If True, Tamm-Dancoff approximation (TDA) is used (linear-response only).',
+    )
```
If this quantity is contingent on linear-response, why not make it part of `type`?
This would save you on sanity-checking logic. If it's for searching, you could make the categories "linear-response TDA" and "linear-response other", or something along those lines.

JosePizarro3 commented on src/nomad_simulations/schema_packages/model_method.py:
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
+        This field describes the effective XC response behaviour assumed in the TDDFT calculation, 
+        whether implemented explicitly (linear-response) or implicitly via the XC potential (real-time).
+          • ALDA / AGGA     : adiabatic local or semi-local
+          • hybrid          : hybrid adiabatic kernel
```
maybe nice to add some references? They can be to code documentation, papers, whatever you think is appropiate to understand the choice

EBB2675 replied:
will do :+1:

ndaelman-hu replied:
there are references in the section-level docs

JosePizarro3 commented on src/nomad_simulations/schema_packages/model_method.py:
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
+        ),
+        description="""
+        Numerical driver:
+          • Casida             : eigenvalue problem in particle-hole space
```
same comment as with the `kernel`

JosePizarro3 commented on src/nomad_simulations/schema_packages/model_method.py:
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
+          • Casida             : eigenvalue problem in particle-hole space
+          • Sternheimer        : frequency-domain Sternheimer
+          • Liouville-Lanczos  : iterative Liouville-Lanczos
+          • propagation        : explicit time evolution (real-time)
```
is `propagation` the only choice if `type="real_time"` and any of the others are for linear response?

JosePizarro3 commented on src/nomad_simulations/schema_packages/model_method.py:
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
+        description='Intended spectral/response target of the TDDFT input.',
+    )
+
+    time_step = Quantity(
```
is this not a variable of the corresponding PhysicalProperty?

ndaelman-hu replied:
I think it fits well here. I don't think the objective is to capture the propagation (that should be added to the workflow level) during the real-time simulation, but just to define a simulation parameter.

EBB2675 committed:
skip default assignments in TDDFT.normalize

JosePizarro3 commented on src/nomad_simulations/schema_packages/model_method.py:
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
+        description='Time integrator used for real-time propagation.',
+    )
+
+    gauge = Quantity(
```
mm what do you mean here?

JosePizarro3 commented on src/nomad_simulations/schema_packages/model_method.py:
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
+        description='Time step used for real-time propagation.',
+    )
+
+    n_steps = Quantity(
```
same comment as for `time_step`

ndaelman-hu replied:
Previous comment accidentally got deleted:

> I think it fits well here. I don't think the objective is to capture the propagation (that should be added to the workflow level) during the real-time simulation, but just to define a simulation parameter.

JosePizarro3 commented on src/nomad_simulations/schema_packages/model_method.py:
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
+        description='Number of time steps in real-time propagation.',
+    )
+
+    propagator = Quantity(
```
same comment as for `time_step`

this is also related with numerical_settings.Mesh (cc @ndaelman-hu)

ndaelman-hu replied:
Yes, I agree that semantically it has more of a numerical settings nature.
@EBB2675 if you add it under a `NumericalSettings` child section, consider adding a reference from here to there.

EBB2675 replied:
will do :+1:

JosePizarro3 reviewed (approved):
All good, just minor stuff

EBB2675 committed:
apply suggestions from Martin

ndaelman-hu commented on src/nomad_simulations/schema_packages/model_method.py:
```diff
@@ -1345,6 +1345,135 @@ class BSE(ExcitedStateMethodology):
+        kernel matches, or to a separate XCFunctional created for the TD kernel
+        when it differs (e.g., LRC/hybrid kernel on top of semi-local ground state).
+        """,
+    )
```
So, the DFT section has to be populated together with this one, correct? Maybe add this to the section's pydocs. (First time reading the description, I thought you were using the reference to describe a workflow, so the extra clarification would be nice.)
I would then also reference to `DFT` instead. Other corrections like Hubbard, Vandermonde, VdW corrections can all still be applied on top.

ndaelman-hu commented on src/nomad_simulations/schema_packages/model_method.py:
```diff
@@ -1345,6 +1345,135 @@ class BSE(ExcitedStateMethodology):
+        if self.type == 'linear_response' and self.is_tamm_dancoff is None:
+            logger.warning(
+                'TDDFT linear_response set but is_tamm_dancoff not specified.'
+            )
```
You aren't covering the other case, where `is_tamm_dancoff` is set, but `type == real-time`.
If the edge cases become sufficiently complex, I would in fact add a test. You can then move this logic out to a dedicated function.

ndaelman-hu reviewed (approved):
Slight restructuring suggestions + few comments to other reviewers

EBB2675 committed:
remove numerical controls from TDDFT

EBB2675 committed:
simplify TDDFT

EBB2675 commented:
- Removed kernel-specific fields from TDDFT (notably `kernel` and `xc_kernel_ref`) to keep TDDFT aligned with the current GW/BSE-style method sections, and to defer XC-kernel details to a **future** workflow-level PR.

- Kept only the essential "flavour" switch via `type = {'linear_response', 'real_time'}` to distinguish frequency-domain response from time-propagation TDDFT.

- Introduced `approximation` = {'full', 'TDA'}, and dropped linear-response-only flags like `is_tamm_dancoff`. In the future, we can enrich this Enum with more approximations.

- Removed real-time "required settings" checks (`time_step`, `n_steps`) since those belong in `numerical_settings.py` and are explicitly postponed in this PR.

- Added light consistency warnings to prevent impossible combinations (e.g., type='real_time' with solver='Casida', or approximation='TDA' with type='real_time').

At the end, it turned into a light method class similar to already existing excited state methodology sections, numerical knobs will come slowly in future PRs :+1: 

Please take a last look :rose: 
@JFRudzinski @ndaelman-hu @JosePizarro3 @kubanmar

EBB2675 committed:
introduce xc, Nathans suggestion
